src/main/java/com/google/devtools/build/lib/skyframe/SequencedSkyframeExecutor.java

```diff
@@ -513,29 +515,40 @@ private void handleDiffsWithMissingDiffInformation(
         diffPackageRootsUnderWhichToCheck.add(pair.getFirst());
       }
 
-      EnumSet<FileType> fileTypesToCheck =
-          EnumSet.of(FileType.EXTERNAL_REPO, FileType.EXTERNAL_IN_MANAGED_DIRECTORY);
+      EnumSet<FileType> fileTypesToCheck = EnumSet.noneOf(FileType.class);
+      Iterable<SkyValueDirtinessChecker> dirtynessCheckers = ImmutableList.of();
+
+      if (!diffPackageRootsUnderWhichToCheck.isEmpty()) {
+        dirtynessCheckers = Iterables.concat(
+            dirtynessCheckers,
+            ImmutableList.of(new MissingDiffDirtinessChecker(diffPackageRootsUnderWhichToCheck)));
+      }
+      if (checkExternalFiles) {
+        dirtynessCheckers = Iterables.concat(dirtynessCheckers, customDirtinessCheckers);
+        fileTypesToCheck =
+            EnumSet.of(FileType.EXTERNAL_REPO, FileType.EXTERNAL_IN_MANAGED_DIRECTORY);
+        if (externalFilesKnowledge.tooManyNonOutputExternalFilesSeen) {
+          fileTypesToCheck.add(FileType.EXTERNAL);
+        }
+      }
       // See the comment for FileType.OUTPUT for why we need to consider output files here.
       if (checkOutputFiles) {
         fileTypesToCheck.add(FileType.OUTPUT);
       }
-      if (externalFilesKnowledge.tooManyNonOutputExternalFilesSeen) {
-        fileTypesToCheck.add(FileType.EXTERNAL);
+      if (checkExternalFiles || checkOutputFiles) {
+        Preconditions.checkArgument(!fileTypesToCheck.isEmpty());
+        dirtynessCheckers = Iterables.concat(
+            dirtynessCheckers,
+            ImmutableList.of(new ExternalDirtinessChecker(tmpExternalFilesHelper, fileTypesToCheck)));
       }
+      Preconditions.checkArgument(!Iterables.isEmpty(dirtynessCheckers));
+
       logger.atInfo().log(
           "About to scan skyframe graph checking for filesystem nodes of types %s",
           Iterables.toString(fileTypesToCheck));
       ImmutableBatchDirtyResult batchDirtyResult;
       try (SilentCloseable c = Profiler.instance().profile("fsvc.getDirtyKeys")) {
-        batchDirtyResult =
-            fsvc.getDirtyKeys(
-                memoizingEvaluator.getValues(),
-                new UnionDirtinessChecker(
-                    Iterables.concat(
-                        customDirtinessCheckers,
-                        ImmutableList.<SkyValueDirtinessChecker>of(
-                            new ExternalDirtinessChecker(tmpExternalFilesHelper, fileTypesToCheck),
-                            new MissingDiffDirtinessChecker(diffPackageRootsUnderWhichToCheck)))));
+        batchDirtyResult = fsvc.getDirtyKeys(memoizingEvaluator.getValues(), new UnionDirtinessChecker(dirtynessCheckers));
```

Comment:
I think it's probably worth taking an ImmutableList#copyOf this dirtynessCheckers list, since I think the iterable gets traversed for every node in the graph.